File: 1D_DP/LC_152.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.12.06

class Solution {
public:
    int maxProduct(vector<int>& nums) {
        // need to consider the prev min and prev max
        int n = nums.size();
        vector<int> dp_min(n, 0), dp_max(n, 0);
        dp_min[0] = nums[0];
        dp_max[0] = nums[0];

        for (int i = 1; i < n; i++){
            dp_max[i] = max(nums[i], max(nums[i] * dp_max[i-1], nums[i] * dp_min[i-1]));
            dp_min[i] = min(nums[i], min(nums[i] * dp_max[i-1], nums[i] * dp_min[i-1]));
        }
        auto it_max = max_element(dp_max.begin(), dp_max.end()), it_max2 = max_element(dp_min.begin(), dp_min.end());
        return max(*it_max, *it_max2);
    }
};

File: 1D_DP/LC_152.py
# Written by: Jian-Yuan Lin
# Date: 2022.12.06

class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        ans = nums[0]
        prev_min, prev_max = nums[0], nums[0]
        n = len(nums)
        for i in range(1, n):
            temp_max = max(nums[i], prev_max * nums[i], prev_min * nums[i])
            prev_min = min(nums[i], prev_max * nums[i], prev_min * nums[i])
            prev_max = temp_max
            ans = max(ans, prev_max)
        
        return ans

        
        
        """
        # need to record the max and min result from previous iteration
        n = len(nums)
        dp_max, dp_min = [0] * n, [0] * n
        dp_max[0] = nums[0]
        dp_min[0] = nums[0]

        for i in range(1, n):
            max_val = max(nums[i] * dp_max[i-1], nums[i] * dp_min[i-1], nums[i])
            min_val = min(nums[i] * dp_max[i-1], nums[i] * dp_min[i-1], nums[i])
            dp_max[i] = max_val
            dp_min[i] = min_val
        return max(max(dp_max), max(dp_min))
        """
        